File: src/stores/user_config.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { invoke } from '@tauri-apps/api/core'

// 用户信息接口
interface User {
    uid: number
    username: string
    avatar: string
}

// 模板配置接口
export interface TemplateConfig {
    copyright: number // 1: 自制, 2: 转载
    source: string
    tid: number // 分区ID
    cover: string // 封面URL
    title: string
    title_prefix: string
    desc: string
    desc_v2?: string
    dynamic: string // 粉丝动?
    subtitle: { open: number; lan: string }
    tag: string // 逗号分隔的标?
    videos: any[]
    dtime?: number // 定时发布时间, 10位时间戳
    open_subtitle: boolean
    interactive: number
    mission_id?: number
    topic_id?: number
    season_id?: number
    section_id?: number
    dolby: number
    lossless_music: number
    no_reprint: number
    open_elec: number
    aid?: number
    up_selection_reply: number
    up_close_reply: number
    up_close_danmu: number
    atomic_int: number
    is_only_self: number
    watermark: number
}

interface UserConfig {
    user: { name: string; cookie: any }
    line?: string
    proxy?: string
    limit: number
    watermark: number
    auto_edit: number
    templates: Record<string, TemplateConfig> // 模板?-> 模板配置
}

// 配置根接?
interface ConfigRoot {
    max_curr: number
    auto_upload: boolean
    auto_start: boolean
    log_level: string
    translation_api_url: string
    translation_api_key: string
    translation_model: string
    translation_prompt: string
    translation_auto: boolean
    config: Record<number, UserConfig> // uid -> 用户配置
}

// 用户模板组合接口
interface UserWithTemplates {
    user: User
    templates: Array<{ name: string; config: TemplateConfig }>
    expanded: boolean // 是否展开
}

interface TemplateCommandResponse {
    success: boolean
    message: string
    template?: TemplateConfig
}

export const useUserConfigStore = defineStore('userConfig', () => {
    const configRoot = ref<ConfigRoot | null>(null)
    const configBase = ref<ConfigRoot | null>(null)
    const loginUsers = ref<User[]>([]) // 洢¼ûб
    const loading = ref(false)
    const error = ref<string | null>(null)

    const userTemplates = computed(() => {
        if (!configRoot.value || loginUsers.value.length === 0) {
            return []
        }

        const result: UserWithTemplates[] = []

        for (const user of loginUsers.value) {
            const userConfig = configRoot.value.config[user.uid]
            const templates =
                userConfig && userConfig.templates
                    ? Object.entries(userConfig.templates).map(([name, config]) => ({
                          name,
                          config
                      }))
                    : []

            result.push({
                user,
                templates,
                expanded: false // 这个状需要单独管?
            })
        }

        return result
    })

    // 用户展开状管?
    const userExpandedState = ref<Record<number, boolean>>({})

    const allUsers = computed(() => userTemplates.value.map(ut => ut.user))
    const totalTemplateCount = computed(() =>
        userTemplates.value.reduce((sum, ut) => sum + ut.templates.length, 0)
    )

    // 获取带有展开状的用户模板列表（用于UI显示?
    const userTemplatesWithExpandedState = computed(() => {
        return userTemplates.value.map(ut => ({
            ...ut,
            expanded: userExpandedState.value[ut.user.uid] || false
        }))
    })

    // 默认模板配置
    const createDefaultTemplate = (): TemplateConfig => ({
        copyright: 1,
        source: '',
        tid: 0,
        cover: '',
        title: '',
        title_prefix: '',
        desc: '',
        desc_v2: undefined,
        dynamic: '',
        subtitle: { open: 0, lan: '' },
        tag: '',
        videos: [],
        dtime: undefined,
        open_subtitle: false,
        interactive: 0,
        mission_id: undefined,
        topic_id: undefined,
        season_id: undefined,
        section_id: undefined,
        dolby: 0,
        lossless_music: 0,
        no_reprint: 0,
        open_elec: 0,
        aid: undefined,
        up_selection_reply: 0,
        up_close_reply: 0,
        up_close_danmu: 0,
        atomic_int: 0,
        is_only_self: 0,
        watermark: 0
    })

    // 配置文件操作
    const loadConfig = async () => {
        loading.value = true
        error.value = null
        try {
            const configData: ConfigRoot = await invoke('load_config')
            configRoot.value = configData
            configBase.value = JSON.parse(JSON.stringify(configData))
            return configData
        } catch (err) {
            error.value = `加载配置失败: ${err}`
            console.error('加载配置失败:', err)
            throw err
        } finally {
            loading.value = false
        }
    }

    const loadBaseConfig = async () => {
        loading.value = true
        error.value = null
        try {
            const configData: ConfigRoot = await invoke('load_config')
            configBase.value = JSON.parse(JSON.stringify(configData))
            return configData
        } catch (err) {
            error.value = `加载配置失败: ${err}`
            console.error('加载配置失败:', err)
            throw err
        } finally {
            loading.value = false
        }
    }

    const saveConfig = async () => {
        loading.value = true
        error.value = null
        try {
            await invoke('save_config', {})
            await loadBaseConfig()
            return true
        } catch (err) {
            error.value = `保存配置失败: ${err}`
            console.error('保存配置失败:', err)
            throw err
        } finally {
            loading.value = false
        }
    }

    // 构建用户模板列表 - 现在只需要设置登录用?
    const buildUserTemplates = async (users: User[]) => {
        // 确保配置已加?
        if (!configRoot.value) {
            await loadConfig()
        }

        // 设置登录用户列表，userTemplates 会自动过计算属更?
        loginUsers.value = users

        return userTemplatesWithExpandedState.value
    }

    // 切换用户展开/收起状?
    const toggleUserExpanded = (userUid: number) => {
        userExpandedState.value[userUid] = !userExpandedState.value[userUid]
    }

    // 获取指定用户的模?
    const getUserTemplates = (userUid: number) => {
        const userTemplate = userTemplates.value.find(ut => ut.user.uid === userUid)
        return userTemplate?.templates || []
    }

    // 获取指定用户的指定模?
    const getUserTemplate = (userUid: number, templateName: string) => {
        const userTemplate = userTemplates.value.find(ut => ut.user.uid === userUid)
        return userTemplate?.templates.find(t => t.name === templateName)?.config
    }

    // 为指定用户添加模?
    const addUserTemplate = async (
        userUid: number,
        templateName: string,
        templateConfig?: TemplateConfig
    ) => {
        // 确保配置已加?
        if (!configRoot.value) {
            await loadConfig()
        }

        if (!configRoot.value) {
            configRoot.value = {
                max_curr: 1,
                auto_start: true,
                auto_upload: true,
                log_level: 'info',
                translation_api_url: '',
                translation_api_key: '',
                translation_model: '',
                translation_prompt: '',
                translation_auto: false,
                config: {}
            }
        }

        // 找到对应的用?
        const user = userTemplates.value.find(ut => ut.user.uid === userUid)?.user
        if (!user) {
            throw new Error('用户不存在')
        }

        // 查找或创建用户配?
        let userConfig = configRoot.value.config[userUid]
        if (!userConfig) {
            throw new Error('用户配置不存在')
        }

        // 棢查模板名是否已存?
        if (userConfig.templates[templateName]) {
            throw new Error('模板名已存在')
        }

        const to_add = templateConfig || createDefaultTemplate()
        to_add.watermark = userConfig.watermark // 使用用户配置中的水印设置
        const server_response: TemplateCommandResponse = await invoke('add_user_template', {
            uid: userUid,
            templateName,
            template: to_add
        })

        if (!server_response || !server_response.success || !server_response.template) {
            throw new Error('添加模板失败: ' + server_response.message)
        }

        // 添加模板
        userConfig.templates[templateName] = server_response.template

        // 保存配置
        await saveConfig()

        return true
    }

    // 删除指定用户的模?
    const removeUserTemplate = async (userUid: number, templateName: string) => {
        if (!configRoot.value) {
            throw new Error('配置未加载')
        }

        const user = userTemplates.value.find(ut => ut.user.uid === userUid)?.user
        if (!user) {
            throw new Error('用户不存在')
        }

        const userConfig = configRoot.value.config[userUid]
        if (!userConfig || !userConfig.templates[templateName]) {
            throw new Error('模板不存在')
        }

        // 删除模板
        delete userConfig.templates[templateName]

        const server_response: TemplateCommandResponse = await invoke('delete_user_template', {
            uid: userUid,
            templateName
        })

        if (!server_response || !server_response.success) {
            throw new Error('删除模板失败: ' + server_response.message)
        }

        // 保存配置
        await saveConfig()

        return true
    }

    // 更新指定用户的模?
    const updateUserTemplate = async (
        userUid: number,
        templateName: string,
        templateConfig: TemplateConfig
    ) => {
        if (!configRoot.value) {
            throw new Error('配置未加载')
        }

        const user = userTemplates.value.find(ut => ut.user.uid === userUid)?.user
        if (!user) {
            throw new Error('用户不存在')
        }

        const userConfig = configRoot.value.config[userUid]
        if (!userConfig || !userConfig.templates[templateName]) {
            throw new Error('模板不存在')
        }

        const server_response: TemplateCommandResponse = await invoke('update_user_template', {
            uid: userUid,
            templateName,
            template: templateConfig
        })

        if (!server_response || !server_response.success || !server_response.template) {
            throw new Error('更新模板失败: ' + server_response.message)
        }

        // 更新模板
        userConfig.templates[templateName] = server_response.template

        // 保存配置
        await saveConfig()

        return true
    }

    // 复制模板
    const duplicateUserTemplate = async (
        userUid: number,
        templateName: string,
        newTemplateName: string
    ) => {
        const templateConfig = getUserTemplate(userUid, templateName)
        if (!templateConfig) {
            throw new Error('源模板不存在')
        }

        return await addUserTemplate(userUid, newTemplateName, {
            ...templateConfig,
            aid: undefined // 复制时清除稿件ID
        })
    }

    // 更新用户基础配置
    const updateUserConfig = async (
        userUid: number,
        updates: Partial<Pick<UserConfig, 'line' | 'proxy' | 'limit' | 'watermark' | 'auto_edit'>>
    ) => {
        if (!configRoot.value) {
            throw new Error('配置未加载')
        }

        const userConfig = configRoot.value.config[userUid]
        if (!userConfig) {
            throw new Error('用户配置不存在')
        }

        // 更新配置
        if ('line' in updates) {
            userConfig.line = updates.line
        }
        if ('proxy' in updates) {
            userConfig.proxy = updates.proxy
        }
        if ('limit' in updates) {
            userConfig.limit = updates.limit!
        }

        if ('watermark' in updates) {
            userConfig.watermark = updates.watermark!
        }

        if ('auto_edit' in updates) {
            userConfig.auto_edit = updates.auto_edit!
        }

        try {
            await invoke('save_user_config', {
                uid: userUid,
                line: userConfig.line,
                proxy: userConfig.proxy,
                limit: userConfig.limit,
                watermark: userConfig.watermark,
                autoEdit: userConfig.auto_edit
            })
            // 保存配置
            await saveConfig()
        } catch (err) {
            throw new Error('保存配置失败: ' + err)
        }

        return true
    }

    const updateGlobalConfig = async (
        updates: Partial<Pick<ConfigRoot, 'max_curr' | 'auto_upload' | 'auto_start' | 'log_level' | 'translation_api_url' | 'translation_api_key' | 'translation_model' | 'translation_prompt' | 'translation_auto'>>
    ) => {
        if (!configRoot.value) {
            throw new Error('配置未加载')
        }

        // 更新全局配置
        Object.assign(configRoot.value, updates)

        try {
            await invoke('save_global_config', {
                maxCurr: configRoot.value.max_curr,
                autoStart: configRoot.value.auto_start,
                autoUpload: configRoot.value.auto_upload,
                logLevel: configRoot.value.log_level,
                translationApiUrl: configRoot.value.translation_api_url,
                translationApiKey: configRoot.value.translation_api_key,
                translationModel: configRoot.value.translation_model,
                translationPrompt: configRoot.value.translation_prompt,
                translationAuto: configRoot.value.translation_auto
            })
            // 保存配置
            await saveConfig()
        } catch (err) {
            throw new Error('保存配置失败: ' + err)
        }

        return true
    }

    return {
        // 状?
        configRoot,
        configBase,
        userTemplates: userTemplatesWithExpandedState, // չ״İ汾
        loading,
        error,

        // 计算属?
        allUsers,
        totalTemplateCount,

        // 方法
        loadConfig,
        saveConfig,
        buildUserTemplates,
        toggleUserExpanded,
        getUserTemplates,
        getUserTemplate,
        addUserTemplate,
        removeUserTemplate,
        updateUserTemplate,
        duplicateUserTemplate,
        updateUserConfig,
        updateGlobalConfig,
        createDefaultTemplate
    }
})

File: src-tauri/src/models/user_config.rs
use anyhow::Result;
use biliup::credential::LoginInfo;
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, fs, path::PathBuf};
use tracing::{debug, info};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserInfo {
    pub uid: u64,
    pub name: String,
    pub cookie: LoginInfo,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Subtitle {
    #[serde(default)]
    pub open: u8,
    #[serde(default)]
    pub lan: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct VideoInfo {
    #[serde(default)]
    pub id: String,
    #[serde(default)]
    pub cid: u64,
    #[serde(default)]
    pub title: String,
    #[serde(default)]
    pub filename: String,
    #[serde(default)]
    pub desc: String,
    #[serde(default)]
    pub path: String,
    #[serde(default)]
    pub finished_at: u64,
    #[serde(default)]
    pub encoding_status: i64,
    #[serde(default)]
    pub status_desc: String,
    #[serde(default)]
    pub group_key: String,
    #[serde(default)]
    pub group_role: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TemplateConfig {
    #[serde(default)]
    pub copyright: u8, // 1: 自制, 2: ת
    #[serde(default)]
    pub source: String,
    #[serde(default)]
    pub tid: u32, // 分区ID
    #[serde(default)]
    pub cover: String, // 封面URL
    #[serde(default)]
    pub title: String,
    #[serde(default)]
    pub title_prefix: String,
    #[serde(default)]
    pub desc: String,
    #[serde(default)]
    pub desc_v2: Option<String>,
    #[serde(default)]
    pub dynamic: String, // ˿̬
    #[serde(default)]
    pub subtitle: Subtitle,
    #[serde(default)]
    pub tag: String, // ŷָıǩ
    #[serde(default)]
    pub videos: Vec<VideoInfo>,
    #[serde(default)]
    pub dtime: Option<u32>, // ʱʱ
    #[serde(default)]
    pub open_subtitle: bool,
    #[serde(default)]
    pub interactive: u8,
    #[serde(default)]
    pub mission_id: Option<u32>,
    #[serde(default)]
    pub topic_id: Option<u32>,
    #[serde(default)]
    pub season_id: Option<u64>,
    #[serde(default)]
    pub section_id: Option<u64>,
    #[serde(default)]
    pub dolby: u8,
    #[serde(default)]
    pub lossless_music: u8, // Hi-Res无损音质
    #[serde(default)]
    pub no_reprint: u8,
    #[serde(default)]
    pub open_elec: u8,
    #[serde(default)]
    pub aid: Option<u64>,
    #[serde(default)]
    pub up_selection_reply: u8,
    #[serde(default)]
    pub up_close_reply: u8,
    #[serde(default)]
    pub up_close_danmu: u8,
    #[serde(default)]
    pub atomic_int: u32,
    #[serde(default)]
    pub is_only_self: u8,
    #[serde(default)]
    pub watermark: u8,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserConfig {
    pub user: UserInfo,
    #[serde(default)]
    pub line: Option<String>,
    #[serde(default)]
    pub proxy: Option<String>,
    #[serde(default)]
    pub limit: u32,
    #[serde(default)]
    pub watermark: u8,
    #[serde(default)]
    pub auto_edit: u8,
    #[serde(default)]
    pub templates: HashMap<String, TemplateConfig>, // ƥconfig.jsonе"templates"ֶ
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConfigRoot {
    #[serde(default)]
    pub max_curr: u32,
    #[serde(default)]
    pub auto_upload: bool,
    #[serde(default)]
    pub auto_start: bool,
    #[serde(default = "default_log_level")]
    pub log_level: String,
    #[serde(default)]
    pub translation_api_url: String,
    #[serde(default)]
    pub translation_api_key: String,
    #[serde(default)]
    pub translation_model: String,
    #[serde(default = "default_translation_prompt")]
    pub translation_prompt: String,
    #[serde(default)]
    pub translation_auto: bool,
    #[serde(default)]
    pub config: HashMap<u64, UserConfig>,
}

fn default_log_level() -> String {
    "info".to_string()
}

fn default_translation_prompt() -> String {
    "You are a professional video title translator. Translate the input title into concise, natural Simplified Chinese. Keep product names, proper nouns, and abbreviations accurate. Output only the translated title without explanation or quotes.".to_string()
}

impl ConfigRoot {
    pub fn from_file(path: &PathBuf) -> Result<Self> {
        let json_content = fs::read_to_string(path)?;
        Ok(serde_json::from_str(&json_content)?)
    }

    pub fn save_to_file(&self, path: &PathBuf) -> Result<()> {
        // ȷĿ¼
        if let Some(parent) = path.parent()
            && !parent.exists()
        {
            fs::create_dir_all(parent)?;
        }

        let json_content = serde_json::to_string_pretty(self)?;
        fs::write(path, json_content)?;
        Ok(())
    }

    pub fn new_user_config(
        &mut self,
        uid: u64,
        username: String,
        cookie: LoginInfo,
        proxy: Option<String>,
    ) -> &Self {
        let user_info = UserInfo {
            uid,
            name: username,
            cookie,
        };
        let user_config = UserConfig {
            user: user_info,
            line: None,
            proxy,
            limit: 0,
            watermark: 0,
            auto_edit: 0,
            templates: HashMap::new(),
        };
        self.config.insert(uid, user_config);

        self
    }

    pub fn add_user_config(&mut self, config: UserConfig) -> &Self {
        let uid = config.user.uid;
        self.config.insert(uid, config);

        self
    }

    pub fn remove_user_config(&mut self, uid: u64) -> Result<&Self> {
        if self.config.remove(&uid).is_some() {
            Ok(self)
        } else {
            Err(anyhow::anyhow!("ûò"))
        }
    }

    pub fn save_user_config(
        &mut self,
        uid: u64,
        line: Option<String>,
        proxy: Option<String>,
        limit: u32,
        watermark: u8,
        auto_edit: u8,
    ) -> Result<&Self> {
        if let Some(user_config) = self.config.get_mut(&uid) {
            info!(
                "UID {} ûø: line={:?}, proxy={:?}, limit={}, watermark={}, auto_edit={}",
                uid, line, proxy, limit, watermark, auto_edit
            );
            user_config.line = line;
            user_config.proxy = proxy;
            user_config.limit = limit;
            user_config.watermark = watermark;
            user_config.auto_edit = auto_edit;
            Ok(self)
        } else {
            Err(anyhow::anyhow!("ûò"))
        }
    }

    pub fn save_global_config(
        &mut self,
        max_curr: u32,
        auto_start: bool,
        auto_upload: bool,
        log_level: String,
        translation_api_url: String,
        translation_api_key: String,
        translation_model: String,
        translation_prompt: String,
        translation_auto: bool,
    ) -> &Self {
        info!(
            "ȫ: max_curr={}, auto_start={}, auto_upload={}, log_level={}",
            max_curr, auto_start, auto_upload, log_level
        );
        self.max_curr = max_curr;
        self.auto_start = auto_start;
        self.auto_upload = auto_upload;
        self.log_level = log_level;
        self.translation_api_url = translation_api_url;
        self.translation_api_key = translation_api_key;
        self.translation_model = translation_model;
        self.translation_prompt = if translation_prompt.trim().is_empty() {
            default_translation_prompt()
        } else {
            translation_prompt
        };
        self.translation_auto = translation_auto;

        self
    }

    pub fn add_user_template(
        &mut self,
        uid: u64,
        template_name: &str,
        template: TemplateConfig,
    ) -> TemplateConfig {
        if let Some(user_config) = self.config.get_mut(&uid) {
            if let Some(old) = user_config
                .templates
                .insert(template_name.to_owned(), template.clone())
            {
                #[cfg(debug_assertions)]
                self.compare_templates(&old, &template);
            }
        }

        template
    }

    pub fn delete_user_template(
        &mut self,
        uid: u64,
        template_name: &str,
    ) -> Option<TemplateConfig> {
        if let Some(user_config) = self.config.get_mut(&uid) {
            user_config.templates.remove(template_name)
        } else {
            None
        }
    }

    pub fn default() -> Self {
        Self {
            max_curr: 1,
            auto_start: true,
            auto_upload: true,
            log_level: default_log_level(),
            translation_api_url: String::new(),
            translation_api_key: String::new(),
            translation_model: String::new(),
            translation_prompt: default_translation_prompt(),
            translation_auto: false,
            config: HashMap::new(),
        }
    }

    #[cfg(debug_assertions)]
    fn compare_templates(&self, old: &TemplateConfig, new: &TemplateConfig) {
        macro_rules! compare_field {
            ($field:ident, $old:expr, $new:expr) => {
                if $old.$field != $new.$field {
                    debug!(
                        "Field '{}' updated: {:?} -> {:?}",
                        stringify!($field),
                        $old.$field,
                        $new.$field
                    );
                }
            };
        }

        fn compare_subtitle(old: &Subtitle, new: &Subtitle) {
            compare_field!(open, old, new);
            compare_field!(lan, old, new);
        }

        fn compare_video_info_vec(old: &Vec<VideoInfo>, new: &Vec<VideoInfo>) {
            // ӡ VideoInfo
            for new_video in new {
                if !old.contains(new_video) {
                    debug!("ƵϢ: {:?}", new_video);
                }
            }
            // ӡɾ VideoInfo
            for old_video in old {
                if !new.contains(old_video) {
                    debug!("ɾƵϢ: {:?}", old_video);
                }
            }
        }

        fn compare_template_fields(old: &TemplateConfig, new: &TemplateConfig) {
            compare_field!(copyright, old, new);
            compare_field!(source, old, new);
            compare_field!(tid, old, new);
            compare_field!(cover, old, new);
            compare_field!(title, old, new);
            compare_field!(title_prefix, old, new);
            compare_field!(desc, old, new);
            compare_field!(desc_v2, old, new);
            compare_field!(dynamic, old, new);
            compare_subtitle(&old.subtitle, &new.subtitle);
            compare_field!(tag, old, new);
            compare_video_info_vec(&old.videos, &new.videos);
            compare_field!(dtime, old, new);
            compare_field!(open_subtitle, old, new);
            compare_field!(interactive, old, new);
            compare_field!(mission_id, old, new);
            compare_field!(dolby, old, new);
            compare_field!(lossless_music, old, new);
            compare_field!(no_reprint, old, new);
            compare_field!(open_elec, old, new);
            compare_field!(aid, old, new);
            compare_field!(up_selection_reply, old, new);
            compare_field!(up_close_reply, old, new);
            compare_field!(up_close_danmu, old, new);
            compare_field!(atomic_int, old, new);
        }

        compare_template_fields(old, new);
    }
}

impl Default for TemplateConfig {
    fn default() -> Self {
        Self {
            copyright: 1, // Ĭ
            source: String::new(),
            tid: 0,
            cover: String::new(),
            title: String::new(),
            title_prefix: String::new(),
            desc: String::new(),
            desc_v2: None,
            dynamic: String::new(),
            subtitle: Subtitle::default(),
            tag: String::new(),
            videos: Vec::new(),
            dtime: None,
            open_subtitle: false,
            interactive: 0,
            mission_id: None,
            topic_id: None,
            season_id: None,
            section_id: None,
            dolby: 0,
            lossless_music: 0,
            no_reprint: 0,
            open_elec: 0,
            aid: None,
            up_selection_reply: 0,
            up_close_reply: 0,
            up_close_danmu: 0,
            atomic_int: 0,
            is_only_self: 0,
            watermark: 0,
        }
    }
}
